### PR TITLE
i#3700: Add IR for AArch64's SIMD LD1/ST1 instructions

### PR DESCRIPTION
**Pull request by AssadHashmi**
These are the multiple single element structure load/stores
from/to one vector register variants:
```
ST1 { <Vt>.<T> }, [<Xn|SP>]
LD1 { <Vt>.<T> }, [<Xn|SP>]
```

### CODE REVIEW

**Comment by AssadHashmi** [comment]
run arm tests

File: core/arch/aarch64/instr_create.h [hunk 1]
```diff
@@ -1796,6 +1796,33 @@
 #define INSTR_CREATE_fnmsub_scalar(dc, Rd, Rm, Rn, Ra) \
     instr_create_1dst_3src(dc, OP_fnmsub, Rd, Rm, Rn, Ra)
 
+/* Advanced SIMD (NEON) memory instructions */
+
+/**
+ * Creates an Advanced SIMD (NEON) LD1 instruction to load multiple
+ * single element structures to one vector register, e.g. LD1 {V0.4H},[X0]
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: nit: missing trailing `.`

File: core/arch/aarch64/instr_create.h [hunk 2]
```diff
@@ -1796,6 +1796,33 @@
...
+ * \param dc      The void * dcontext used to allocate memory for the instr_t.
+ * \param q       The destination vector register operand.
+ * \param r       The source memory operand.
+ * \param s       The size of the vector element.
+ */
+#define INSTR_CREATE_ld1_multi_1(dc, q, r, s) instr_create_1dst_2src(dc, OP_ld1, q, r, s)
+
+/**
+ * Creates an Advanced SIMD (NEON) ST1 instruction to store multiple
+ * single element structures from one vector register, e.g. ST1 {V1.2S},[X1]
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: nit: missing trailing `.`

File: core/arch/aarch64/instr_create.h [hunk 3]
```diff
@@ -1796,6 +1796,33 @@
...
+ * \param r       The destination memory operand.
+ * \param q       The source vector register operand.
+ * \param s       The size of the vector element.
+ */
+#define INSTR_CREATE_st1_multi_1(dc, r, q, s) instr_create_1dst_2src(dc, OP_st1, r, q, s)
+
+/* TODO: Remaining advanced SIMD (NEON) memory instructions:
+#define INSTR_CREATE_ld2/3/4_multi_2/3/4()
+#define INSTR_CREATE_ld1/2/3/4_single()
+and st1 equivalents including post-index variants */
```
Comment (derekbruening) [inline core/arch/aarch64/instr_create.h]: style: Add the issue number, so `TODO i#2626: `.

Also we prefer `*/` on its own line.

**Comment by fhahn** [comment]
Great, thanks Assad! Did you use the scripts to auto-generate them?

File: suite/tests/api/ir_aarch64.c [hunk 4]
```diff
@@ -2480,6 +2480,142 @@ test_asimdsame(void *dc)
...
+        opnd_create_base_disp_aarch64(DR_REG_X0, DR_REG_NULL, 0, false, 0, 0, OPSZ_2),
+        OPND_CREATE_DOUBLE());
+    test_instr_encoding(dc, OP_ld1, instr);
+
+    /* Store multiple 1-element structures (to 1, 2, 3 or 4 registers)
+       Naming convention based on official mnemonics:
+       INSTR_CREATE_st1_multi_<n>() where <n> is 1, 2, 3 or 4
+
+       ST1 { <Vt>.<T> }, [<Xn|SP>]
+       ST1 { <Vt>.<T>, <Vt2>.<T> }, [<Xn|SP>]
```
Comment (fhahn) [inline suite/tests/api/ir_aarch64.c]: Currently, only the 1 operand version is tested, right?

File: suite/tests/api/ir_aarch64.c [hunk 5]
```diff
@@ -2480,6 +2480,142 @@ test_asimdsame(void *dc)
...
+    instr_t *instr;
+
+    /* Advanced SIMD memory (multiple structures) */
+
+    /* Load multiple 1-element structures (to 1, 2, 3 or 4 registers)
+       Naming convention based on official mnemonics:
+       INSTR_CREATE_ld1_multi_<n>() where <n> is 1, 2, 3 or 4
+
+       LD1 { <Vt>.<T> }, [<Xn|SP>]
+       LD1 { <Vt>.<T>, <Vt2>.<T> }, [<Xn|SP>]
```
Comment (fhahn) [inline suite/tests/api/ir_aarch64.c]: Currently, only the 1 operand version is tested, right?

File: core/arch/aarch64/codec.txt [hunk 6]
```diff
@@ -756,7 +756,7 @@ x0110111xxxxxxxxxxxxxxxxxxxxxxxx  tbnz   tbz
 0x001100000000000010xxxxxxxxxxxx  st1    memvm : vmsz vt0 vt1 vt2 vt3
 0x001100000000000100xxxxxxxxxxxx  st3    memvm : vmsz vt0 vt1 vt2
 0x001100000000000110xxxxxxxxxxxx  st1    memvm : vmsz vt0 vt1 vt2
-0x001100000000000111xxxxxxxxxxxx  st1    memvm : vmsz vt0
+0x001100000000000111xxxxxxxxxxxx  st1    memvm : vt0 vmsz
```
Comment (fhahn) [inline core/arch/aarch64/codec.txt]: Did we flip the position of the vmsz operand here to be more in line with the XXX_sz operands, which come last?

**Comment by AssadHashmi** [comment]
> Just some style nits. I did not look in the manual to check the encoding bits. Are there tests, either automated or manual pre-commit, to ensure this is the right encoding by comparing to some other decoder?

In this case I've tested the instruction macros in a client which copies vectors between buffers. It's an application specific client. At some point I'll port it to a client targeted at testing which is safe to be upstreamed as part of the test suite. It can be the basis of further AArch64 SIMD work.

File: suite/tests/api/ir_aarch64.c (hunk 4 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Yes. Hopefully more will follow as time allows.

File: suite/tests/api/ir_aarch64.c (hunk 5 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Yes. Hopefully more will follow as time allows.

File: core/arch/aarch64/codec.txt (hunk 6 above)
Comment (AssadHashmi) [reply core/arch/aarch64/codec.txt]: Yes

**Comment by AssadHashmi** [comment]
> Great, thanks Assad! Did you use the scripts to auto-generate them?

No, unfortunately. There's still some issues and work required and resourcing is . . . variable ;-)

File: suite/tests/api/ir_aarch64.c (hunk 4 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Hmm...good spot. Perhaps I should remove those comments as they could be misleading?

File: suite/tests/api/ir_aarch64.c (hunk 5 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Hmm...good spot. Perhaps I should remove those comments as they could be misleading?

File: suite/tests/api/ir_aarch64.c (hunk 4 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Maybe not. The header comments are useful placeholders and each test call is explicitly commented to make it clear it's the 1 operand version being run.

File: suite/tests/api/ir_aarch64.c (hunk 5 above)
Comment (AssadHashmi) [reply suite/tests/api/ir_aarch64.c]: Maybe not. The header comments are useful placeholders and each test call is explicitly commented to make it clear it's the 1 operand version being run.

File: suite/tests/api/dis-a64.txt [hunk 7]
```diff
@@ -74,7 +74,7 @@
 0b9f13ff : add    wzr, wzr, wzr, asr #4   : add    %wzr %wzr asr $0x04 -> %wzr
 0c0007ff : st4    {v31.4h, v0.4h, v1.4h, v2.4h}, [sp]: st4    $0x01 %d31 %d0 %d1 %d2 -> (%sp)[32byte]
 0c0067ff : st1    {v31.4h, v0.4h, v1.4h}, [sp]: st1    $0x01 %d31 %d0 %d1 -> (%sp)[24byte]
-0c0077ff : st1    {v31.4h}, [sp]          : st1    $0x01 %d31 -> (%sp)[8byte]
+0c0077ff : st1    {v31.4h}, [sp]          : st1    %d31 $0x01 -> (%sp)[8byte]
```
Comment (fhahn) [inline suite/tests/api/dis-a64.txt]: Would it be possible to add tests for all sizes? same for ld1.

File: core/arch/aarch64/instr.c [hunk 8]
```diff
@@ -335,6 +335,12 @@ reg_is_gpr(reg_id_t reg)
     return (DR_REG_X0 <= reg && reg <= DR_REG_WSP);
 }
 
+bool
+reg_is_simd(reg_id_t reg)
```
Comment (fhahn) [inline core/arch/aarch64/instr.c]: unused?

File: core/arch/aarch64/codec.txt (hunk 6 above)
Comment (fhahn) [reply core/arch/aarch64/codec.txt]: Makes sense! Is there a reason you did not change it for ld1?

Comment (AssadHashmi) [reply core/arch/aarch64/codec.txt]: > Makes sense! Is there a reason you did not change it for ld1?

ld1 doesn't need to change as its fields are already correctly positioned w.r.t. encoding string:
```
0x001100010000000111xxxxxxxxxxxx  ld1    vt0 : memvm vmsz
```

File: core/arch/aarch64/instr.c (hunk 8 above)
Comment (AssadHashmi) [reply core/arch/aarch64/instr.c]: I added `reg_is_simd()` in order for AArch64 to conform to the API.

File: core/arch/aarch64/codec.txt (hunk 6 above)
Comment (fhahn) [reply core/arch/aarch64/codec.txt]: Right!

File: core/arch/aarch64/instr.c (hunk 8 above)
Comment (fhahn) [reply core/arch/aarch64/instr.c]: I see. Would it be possible to add a test for it to https://github.com/DynamoRIO/dynamorio/blob/master/suite/tests/api/opnd-a64.c ?

File: core/arch/aarch64/instr.c [hunk 9]
```diff
@@ -335,6 +335,12 @@ reg_is_gpr(reg_id_t reg)
...
+{
+    return (DR_REG_Q0 <= reg && reg <= DR_REG_B31);
```
Comment (fhahn) [inline core/arch/aarch64/instr.c]: I am a bit surprised that `S`, `H` and `B` registers are included here. I thought the only vector registers used by instructions are the 128 bit `Q`, as in `v8.2d, v8.4s, v8.8h` and 64 bit `D` as in `v8.2s`, `v8.4h`. Am I missing something and are there encodings that use `S`, `H` or `B` registers as vector register?

Comment (AssadHashmi) [reply core/arch/aarch64/instr.c]: My (simple) understanding is that all `reg_is_simd()` is saying is whether a given register is SIMD or not, and `B`/`H`/`S`/`D` are components of a SIMD `Q` register. That's independent of any encoding issues. Have I missed something?

**Comment by derekbruening** [comment]
Remember the 'Fixes #NNN' or 'Issue: #NNN' for future commits: this one had neither and so has no link to the issue.

Can #3700 be closed now?

**Comment by AssadHashmi** [comment]
> Remember the 'Fixes #NNN' or 'Issue: #NNN' for future commits: this one had neither and so has no link to the issue.
Gah! My bad.

> Can #3700 be closed now?

**Comment by derekbruening** [comment]
> > Remember the 'Fixes #NNN' or 'Issue: #NNN' for future commits: this one had neither and so has no link to the issue.
> 
> Gah! My bad.

If you've run the devsetup script, it sets a commit message template which makes it easier to remember by pre-populated each commit message with the suggested form.

**Pull request merged**